brick/dockerlib.py

```diff
@@ -12,6 +14,38 @@
 docker_client = docker.from_env()
 
 
+def is_built_up_to_date(tag: str, dependency_paths: List[str]) -> bool:
```

Comment:
First up, thanks for actually doing this work now - and let me know if you'd prefer to wait with continuing work on it until you're actually in the office! 😄 

---

As far as I can tell, this way of checking for dependencies doesn't support listing full directories and using globbing - it only checks when the actual directory was modified and not the content 🤔 

Here's a few examples of how dependencies are being listed:

```bash
	inputs:
	  - src
      - ../../../{co2engine,utils}
      - ../../../../northapp/src/errorUtils.js
      - ../../../../northapp/src/schemas/
      - gatsby-*.js
      - ./*.py
      - ./{.babelrc,babel.config.js,.eslintignore}
```

Thirdly, is looping through dependencies and checking file/folder timestamps for each the best way of checking for changes? I would imagine it could get quite intense for some packages, although maybe it's speedy enough to not be a problem (fx if someone accidentally added `node_modules`)?

I remember previously looking into using `git` for listing the files changed in a branch, but can't remember why I stopped/never took it further 🙈 

For inspiration, [baur](https://github.com/simplesurance/baur) (another build tool) does it like this: 
> baur calculates a SHA384 digest for all inputs and stores the digest in the database when an application was built and its artifacts uploaded (baur build). The digest is used to detect if a previous build for the same input files exists. If a build exist, the application does not need to be rebuilt, otherwise a build is done.

Reply:
> I remember previously looking into using git for listing the files changed in a branch, but can't remember why I stopped/never took it further 🙈

Jumping in here to refresh our memory: It's probably because you might want to re-build even though nothing changed in git (for example if the last build crashed).

Reply:
> As far as I can tell, this way of checking for dependencies doesn't support listing full directories and using globbing - it only checks when the actual directory was modified and not the content 🤔

Globbing is supported here as the dependency_paths are already expanded. But I did just check, and any changes in a folder does not reflect the modified timestamp. So we need to recursively check folders for changes.

> Thirdly, is looping through dependencies and checking file/folder timestamps for each the best way of checking for changes? I would imagine it could get quite intense for some packages, although maybe it's speedy enough to not be a problem (fx if someone accidentally added node_modules)?

File access is really fast, so I don't think it should be an issue. But let us see once the recursive traversal of folders is implemented.

Reply:
@madsnedergaard do you have some timings for trying this out on CI? And what about locally, do you see similar speed ups?

Reply:
> First up, thanks for actually doing this work now - and let me know if you'd prefer to wait with continuing work on it until you're actually in the office! 😄

It was fun! And it was nice to wrap it up in a PR. Let us see how much time I have, else we will wrap it up at a later point. ;)

Reply:
Yeah, no change in timings - it's building every package.

But I guess it's because files are created from fresh in CI on each build, so file timestamps is an issue:

![Screenshot 2020-08-25 at 15 32 18](https://user-images.githubusercontent.com/3296643/91180591-2ffafb00-e6e8-11ea-8378-70635bd688cf.png)

https://drone.tmrow.com/tmrowco/tmrow/5082/1/3

Reply:
> But I guess it's because files are created from fresh in CI on each build, so file timestamps is an issue:

Auch. I missed the fact that file time stamps are not preserved in Git. Then my approach here is flawed... There are script to update file time stamps based on git history, but I don’t like to rely on that. 

I’ll probably pull out the extract optimization into a separate PR. We could probably also do some more optimizations here. 

And we should debug why build kit is that slow at diffing.